### PR TITLE
[Master] Fix incorrect results for panic in `module-const-declaration` with arithmetic operators

### LINKED ISSUE #36648: Incorrect result for multiplication with out of range int
**Description:**
$title

**Steps to reproduce:**
```
const num1 = -2;
const  num2 = -9223372036854775807;
const int ans = num1 * num2; // ans = -2
```

### LINKED ISSUE #37711: [Bug]: Remainder (%) operand is not supported for `module-const-decl`
### Description

$title

### Steps to Reproduce

```
const num1 = 2;
const num2 = 5;
const int ans = num2 % num1;
```

This gives compilation error
```
ERROR [test_3.bal:(3:17,3:28)] const expressions are not yet supported here
error: compilation contains errors
```
But according to the spec, it is supported

### Affected Version(s)

Ballerina 2201.1.1 (Swan Lake Update 1)
Language specification 2022R2

### OS, DB, other environment details and versions

_No response_

### Related area

Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by prakanth97**
## Purpose

Fixes #36648
Fixes #37711

## Approach
Give compile-time errors for panic scenarios instead of giving wrong output.

## Samples
```
const NUM1 = -1;
const int NUM2 = -9223372036854775807 - 1;
const int NUM3 = 0;
const int ANS1 = NUM2 / NUM1;          // error
const int ANS2 = NUM2 / NUM3;          // error

const int ANS3 = NUM2 * NUM1;          // error
const int ANS4 = NUM2 * 5;             // error

const int ANS5 = NUM2 % NUM3;          // error
const int ANS6 = -NUM2;                // error

const int ANS7 = NUM2 + NUM1;          // error
const int ANS8 = NUM2 - 1;             // error
```
Now, these test cases will give compile-time error

## Remarks
https://github.com/ballerina-platform/ballerina-lang/issues/13944

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by prakanth97** [comment]
While we are logging the java errors for constant-expr https://github.com/ballerina-platform/ballerina-lang/blob/68a7396966b4654a500ba7fbd2df92ccb923105b/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java#L333, 

we stick with the java like error for 
```
const NUM1 = -1;
const int NUM2 = -9223372036854775807 - 1;
const int ANS1 = NUM2 / NUM1;          // error
```
which is `invalid constant expression, reason 'long overflow'`. Is that correct or do we need to have custom compile-time error codes?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java [hunk 1]
```diff
@@ -596,7 +607,7 @@ private void checkUniqueness(BLangConstant constant) {
                     dlog.error(constant.name.pos, DiagnosticErrorCode.ALREADY_INITIALIZED_SYMBOL, nameString);
                 } else {
                     BLangConstantValue lastValue = constantMap.get(nameString);
-                    if (!value.equals(lastValue)) {
+                    if (!value.toString().equals(lastValue.toString())) {
```
Comment (prakanth97) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java:610]: Made this change to fix CCE for 
```
const NUM1 = -1;
const int NUM2 = -9223372036854775807 - 1;
const int ANS7 = NUM2 - 1; 
const int ANS7 = NUM2 + NUM1; 
```

Is that acceptable?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java [hunk 2]
```diff
@@ -376,12 +378,11 @@ private BLangConstantValue calculateBitWiseOp(BLangConstantValue lhs, BLangConst
     }
 
     private BLangConstantValue calculateAddition(BLangConstantValue lhs, BLangConstantValue rhs) {
-        // TODO : Handle overflow in numeric values.
         Object result = null;
         switch (Types.getReferredType(this.currentConstSymbol.type).tag) {
             case TypeTags.INT:
             case TypeTags.BYTE: // Byte will be a compiler error.
-                result = (Long) lhs.value + (Long) rhs.value;
+                result = Math.addExact((Long) lhs.value, (Long) rhs.value);
```
Comment (SandaruJayawardana) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java:385]: Here, we have to handle `byte` case separately. Please check the other operations also.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java (hunk 1 above)
Comment (SandaruJayawardana) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java:610]: This change causes to allow invalid constant declarations like following example.

```ballerina
public enum LiftStatus {
    CLOSED = "123"
}

const CLOSED = 123; // No errors!
```

**Comment by SandaruJayawardana** [comment]
> we stick with the java like error for
    const NUM1 = -1;
    const int NUM2 = -9223372036854775807 - 1;
    const int ANS1 = NUM2 / NUM1;          // error
which is invalid constant expression, reason 'long overflow'. Is that correct or do we need to have custom compile-time error codes?

IMO, we have to change the error messages since Ballerina has different data types compared to Java. For example, `long` data type in above error msg (`long overflow`).

File: tests/jballerina-unit-test/src/test/resources/test-src/types/constant/simple-literal-constant-negative.bal [hunk 3]
```diff
@@ -321,3 +321,19 @@ Ints2 ints2 = 2;
 
 float:NaN floatNan = 1.0;
 float:Infinity floatInf = 1.0;
+
+// Test arithmatic errors
+const NUM1 = -1;
```
Comment (lasinicl) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/constant/simple-literal-constant-negative.bal:326]: Shall we add more tests with different combinations of constants, when both negative, both positive, one negative and one positive etc...?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java [hunk 4]
```diff
@@ -487,6 +489,9 @@ private BLangConstantValue calculateMod(BLangConstantValue lhs, BLangConstantVal
     }
 
     private Object calculateNegationForInt(BLangConstantValue value) {
+        if ((Long) (value.value) == Long.MIN_VALUE) {
```
Comment (lasinicl) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java:492]: Do we have a test to cover this change?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java (hunk 2 above)
Comment (SandaruJayawardana) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java:385]: Since byte scenario is not yet allowed, we can go with this change. Byte scenario will be fixed with https://github.com/ballerina-platform/ballerina-lang/issues/36538.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java (hunk 1 above)
Comment (prakanth97) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java]: Updated b3fb7aaad3e156c54a1388007bd0154774194e42

**Comment by prakanth97** [comment]
> > we stick with the java like error for
> > const NUM1 = -1;
> > const int NUM2 = -9223372036854775807 - 1;
> > const int ANS1 = NUM2 / NUM1;          // error
> > which is invalid constant expression, reason 'long overflow'. Is that correct or do we need to have custom compile-time error codes?
> 
> IMO, we have to change the error messages since Ballerina has different data types compared to Java. For example, `long` data type in above error msg (`long overflow`).
Updated 60870f0bfefe797664b814cafceb5590e491915c

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java (hunk 4 above)
Comment (prakanth97) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java:508]: Yes https://github.com/ballerina-platform/ballerina-lang/pull/37726/files#diff-f7b80acafb6a5c9b42be1aea1acfa78b59911c59f6a479ac5cc5265b47e1a213R336

File: tests/jballerina-unit-test/src/test/resources/test-src/types/constant/simple-literal-constant-negative.bal (hunk 3 above)
Comment (prakanth97) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/constant/simple-literal-constant-negative.bal:326]: Added 526aee55dfdb85ca0d7e62171ca467a9584f6a48

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java [hunk 5]
```diff
@@ -311,6 +311,8 @@ private BLangConstantValue calculateConstValue(BLangConstantValue lhs, BLangCons
                     return calculateMultiplication(lhs, rhs);
                 case DIV:
                     return calculateDivision(lhs, rhs);
+                case MOD:
```
Comment (SandaruJayawardana) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java:314]: Shall we add more test cases to cover the spec related to remainder operator (to cover edge cases mentioned in the spec) https://ballerina.io/spec/lang/master/#section_6.23

Comment (prakanth97) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java:314]: Sure.

**Comment by SandaruJayawardana** [comment]
We need to handle overflow scenarios of `decimal` also. 
e.x. Following code crashes the compiler.
```ballerina
const decimal X = 1e100000000000000050 * 1e10000000000300050;
```

File: compiler/ballerina-lang/src/main/resources/compiler.properties [hunk 6]
```diff
@@ -1919,3 +1919,6 @@ error.unsupported.computed.resource.access.path.segment.type=\
 error.unsupported.resource.access.rest.segment.type=\
     unsupported resource access rest segment type: expected array of ''int'', ''string'', ''float'', ''boolean'',\
    ''decimal'' but found ''{0}''
+
+error.int.range.overflow=\
+  'int' range overflow
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/resources/compiler.properties:1924]: If we mention the supported limits, it will be useful for the user

**Comment by prakanth97** [comment]
> We need to handle overflow scenarios of `decimal` also. e.x. Following code crashes the compiler.
> 
> ```ballerina
> const decimal X = 1e100000000000000050 * 1e10000000000300050;
> ```

Fix for this issue is done here https://github.com/ballerina-platform/ballerina-lang/pull/37823

File: compiler/ballerina-lang/src/main/resources/compiler.properties (hunk 6 above)
Comment (prakanth97) [reply compiler/ballerina-lang/src/main/resources/compiler.properties]: Updated 91289db7c7c17982b7483294300408a9fd1bb23b

**Comment by SandaruJayawardana** [comment]
> We need to handle overflow scenarios of decimal also. e.x. Following code crashes the compiler.
    const decimal X = 1e100000000000000050 * 1e10000000000300050;
>>Fixes in https://github.com/ballerina-platform/ballerina-lang/pull/37823

Once https://github.com/ballerina-platform/ballerina-lang/pull/37823 is merged, we may have to handle the constant overflow separately.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/BLangConstantValue.java [hunk 7]
```diff
@@ -56,6 +56,10 @@ public boolean equals(Object o) {
                 return false;
             }
 
+            if (this.value == null) {
+                return false;
```
Comment (SandaruJayawardana) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/BLangConstantValue.java:60]: Shall we add a test case to cover this?

**Comment by prakanth97** [comment]
> > We need to handle overflow scenarios of decimal also. e.x. Following code crashes the compiler.
> > const decimal X = 1e100000000000000050 * 1e10000000000300050;
> > > Fixes in #37823
> 
> Once #37823 is merged, we may have to handle the constant overflow separately.

According to the spec https://ballerina.io/spec/lang/master/#section_6.23,
Decimal overflow panic will not happen, but compiler crashes because java can not hold the larger decimal value as mentioned in https://github.com/ballerina-platform/ballerina-lang/pull/37823. 

For this example,
```
const decimal X = 1e100000000000000050 * 1e10000000000300050;
```
The fix will give compile-time error.  But, 
```
const decimal ANS8 = 100 % 9.999999999999999999999999999999999e6144;
``` 
This is not compile-time error or panic.

```
const decimal a = 5 % 0;
const decimal a = 0 % 0;
``` 
These are panic scenarios

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/BLangConstantValue.java (hunk 7 above)
Comment (prakanth97) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/BLangConstantValue.java:60]: Sure, there should be a negative test

**Comment by pcnfernando** [comment]
@prakanth97 can you please resolve the conflict

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java [hunk 8]
```diff
@@ -405,7 +412,12 @@ private BLangConstantValue calculateSubtract(BLangConstantValue lhs, BLangConsta
         switch (Types.getReferredType(this.currentConstSymbol.type).tag) {
             case TypeTags.INT:
             case TypeTags.BYTE: // Byte will be a compiler error.
-                result = (Long) lhs.value - (Long) rhs.value;
+                try {
+                    result = Math.subtractExact((Long) lhs.value, (Long) rhs.value);
+                } catch (ArithmeticException ae) {
+                    dlog.error(currentPos, DiagnosticErrorCode.INT_RANGE_OVERFLOW_ERROR);
+                    return new BLangConstantValue(null, this.currentConstSymbol.type);
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java:419]: We don't have to return here. At this level, the result will be null; if we break in the following line, the last statement will return the null const value. Check other places as well

Comment (prakanth97) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java]: Updated  9c8454389b2969a3679513c74c57a8390d137c18

**Comment by prakanth97** [comment]
> @prakanth97 please check on the build

Build failure is not due to my fix, it is resolved in the master

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 9]
```diff
@@ -4050,6 +4055,14 @@ boolean isSigned32LiteralValue(Long longObject) {
         return (longObject >= SIGNED32_MIN_VALUE && longObject <= SIGNED32_MAX_VALUE);
     }
 
+    BigDecimal getValidDecimalNumber(Location pos, BigDecimal bd) {
+        if (bd.compareTo(DECIMAL_MAX) > 0 || bd.compareTo(DECIMAL_MIN) < 0) {
+            dlog.error(pos, DiagnosticErrorCode.OUT_OF_RANGE, bd.toString(), symTable.decimalType);
+            return null;
+        }
```
Comment (lasinicl) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:4062]: Need to handle decimal subnormals too since decimal subnormals should be treated as zero as per the spec. 
https://github.com/ballerina-platform/ballerina-lang/blob/master/bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/DecimalValue.java#L102

File: tests/jballerina-unit-test/src/test/resources/test-src/types/constant/simple-literal-constant-negative.bal [hunk 10]
```diff
@@ -321,3 +321,35 @@ Ints2 ints2 = 2;
 
 float:NaN floatNan = 1.0;
 float:Infinity floatInf = 1.0;
+
+// Test arithmatic errors
```
Comment (lasinicl) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/constant/simple-literal-constant-negative.bal:325]:
```suggestion
// Test arithmetic errors
```

Comment (lasinicl) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/constant/simple-literal-constant-negative.bal:325]: We can add this comment in line no#331 and get rid of `//error` comment in the lines which have tested the arithmetic operations

File: tests/jballerina-unit-test/src/test/resources/test-src/types/constant/simple-literal-constant-negative.bal [hunk 11]
```diff
@@ -321,3 +321,35 @@ Ints2 ints2 = 2;
...
+
+const decimal ANS12 = 5 % 0;           // error
+
+const int ANS7 = NUM2 - 1;             // error
+
+const decimal d1 = 9.999999999999999999999999999999999E6001d * 1E145d;   // error
+const decimal d2 = -9.999999999999999999999999999999999E6141d * 1E5d;    // error
+const decimal d3 = 9.999999999999999999999999999999999E6144d + 1E6143d;  // error
+const decimal d4 = -1E6144d - 9.999999999999999999999999999999999E6144d; // error
+const decimal d5 = 1E614d / 2E-5800d;                                    // error
```
Comment (lasinicl) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/constant/simple-literal-constant-negative.bal:355]: Shall we add some tests for unary operators as well?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java (hunk 9 above)
Comment (prakanth97) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:4062]: Need to revert this d254b924636685793772e40e65142639d6130e12

File: tests/jballerina-unit-test/src/test/resources/test-src/types/constant/simple-literal-constant-negative.bal (hunk 11 above)
Comment (prakanth97) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/constant/simple-literal-constant-negative.bal]: Added 9d9f49229743a0d2c9229b098cc2bde7a8d31c25

There is an existing test as well https://github.com/ballerina-platform/ballerina-lang/pull/37726/files#diff-f7b80acafb6a5c9b42be1aea1acfa78b59911c59f6a479ac5cc5265b47e1a213R339

**Review by lasinicl — APPROVED** [review]

**Comment by SandaruJayawardana** [comment]
I am getting `0` for following example. Is it correct? @prakanth97 
```
const decimal d5 = 2E-5800d * 2E-5800d * 2E-5800d  * 1E-614d  * 2E-5800d;
```

**Review by SandaruJayawardana — APPROVED** [review]

**Review by pcnfernando — APPROVED** [review]

**Comment by pcnfernando** [comment]
@prakanth97 can you check on the pipeline failures
